libopenstorage/operator
# PWX-36551: Update plugin image with change in versions if autoupdatecomponents is used

**What this PR does / why we need it**: Previously, the image versions for dynamic plugin and dynamic plugin proxy where not updated when versions configmap / URL endpoint was changed even though autoUpdateComponents was set to Once or Always. This PR makes sure the images of both get updated in such a case , i.e; when autoUpdateComponents is Once/Always

**Which issue(s) this PR fixes** (optional)
Closes # PWX-36551

**Testing notes** :

1.  Using versions configmap

 a) When autoUpdateComponents is set to Once
   - Create versions configmap using `kubectl -n kube-system create configmap px-versions --from-file=versions.yaml`. 
       The versions.yaml file should contain the component image: `dynamicPluginProxy: nginxinc/nginx-unprivileged:1.23`
   - Create a storage cluster with any portworx version
   - Verify with the stc yaml , desired images field should contain the image `dynamicPluginProxy: nginxinc/nginx- 
        unprivileged:1.23`
   - Edit the stc and add the field - `autoUpdateComponents: Once` and save. Wait till the change is applied and px pods 
      retsrart
   - Edit the px-versions cm and change image of dynamicPluginProxy to `dynamicPluginProxy: nginxinc/nginx- 
      unprivileged:1.25`.
   - Wait for a minute and check the stc yaml. This time the desired images of dynamicPluginProxy should be 1.25 instead 
      of 1.23

 b) When autoUpdateComponents is set to Always
   - Create versions cm as before
   - Create a storage cluster with field - `autoUpdateComponents: Always`
   - Once the stc is running, check the yaml to verify desired images has  `dynamicPluginProxy: nginxinc/nginx- 
       unprivileged:1.23`
   - Edit the px-versions cm and change image of dynamicPluginProxy to `dynamicPluginProxy: nginxinc/nginx- 
         unprivileged:1.25`.
   - Verify yaml of stc has desired images `dynamicPluginProxy: nginxinc/nginx-unprivileged:1.25`.

2. Using URL endpoint for versions
 
 a) When autoUpdateComponents is set to Once
  - Create a http server which hosts a custom px-versions file in an endpoint. This versions file should use the older 
     version-  `dynamicPluginProxy: nginxinc/nginx-unprivileged:1.23`
  - Create a stc with env variable 
   `- name: PX_RELEASE_MANIFEST_URL
      value: "http://<IP>:<PORT>"`
     where the value will be the IP and port of the endpoint where the custom px-versions file is hosted.
 - Verify stc yaml to older image for dynamicPluginProxy
 - Add the `autoUpdateComponents: Once` field to the stc
 - Edit the value of PX_RELEASE_MANIFEST_URL env to - "https://edge-install.portworx.com/3.1.1/version"
 - Wait for px pods to restart, check yaml to verify image version for dynamicPluginProxy is 1.25 in desired images
 
  b) When autoUpdateComponents is set to Always
  - Create the http server as before
  - Create an stc with the following env variables:

```
  - name: PX_RELEASE_MANIFEST_REFRESH_INTERVAL_MINS
      value: "1"
   - name: PX_RELEASE_MANIFEST_URL
      value: "http:// <IP> : <PORT> " 
```
  - Also add the field `autoUpdateComponents: Always` to the stc 
  - Verify stc yaml to older image for dynamicPluginProxy
  -  Edit the value of PX_RELEASE_MANIFEST_URL env to - "https://edge-install.portworx.com/3.1.1/version"
  - After a while verify image version for dynamicPluginProxy is 1.25 in desired images
